lms/static/scripts/frontend_apps/components/test/SubmitGradeForm-test.js

```diff
@@ -325,6 +328,134 @@ describe('SubmitGradeForm', () => {
+
+      // Clicking the toggle will display the popover
+      togglePopover(wrapper);
+      assert.isTrue(commentPopoverExists(wrapper));
+      assert.isTrue(getToggleButton(wrapper).prop('expanded'));
+
+      // A second click will hide the popover
+      togglePopover(wrapper);
+      assert.isFalse(commentPopoverExists(wrapper));
+      assert.isFalse(getToggleButton(wrapper).prop('expanded'));
+    });
+
+    it('hides the popover when `Escape` is pressed', () => {
+      const wrapper = renderForm({ acceptComments: true });
+
+      // Show popover
+      togglePopover(wrapper);
+      assert.isTrue(commentPopoverExists(wrapper));
+
+      document.body.dispatchEvent(
+        new KeyboardEvent('keydown', { key: 'Escape' })
+      );
+      wrapper.update();
+
+      assert.isFalse(commentPopoverExists(wrapper));
+    });
+
+    it('hides the popover when clicking away', () => {
+      const wrapper = renderForm({ acceptComments: true });
+
+      // Show popover
+      togglePopover(wrapper);
+      assert.isTrue(commentPopoverExists(wrapper));
+
+      const externalButton = document.createElement('button');
+      document.body.append(externalButton);
+      externalButton.click();
+      wrapper.update();
+
+      assert.isFalse(commentPopoverExists(wrapper));
+
+      externalButton.remove();
+    });
+
+    it('adds proper title to toggle button based on the existence of a comment', async () => {
+      const wrapper = renderForm({ acceptComments: true });
+
+      assert.equal(getToggleButton(wrapper).prop('title'), 'Add comment');
+      await waitForGradeFetch(wrapper);
+      assert.equal(getToggleButton(wrapper).prop('title'), 'Edit comment');
+    });
+
+    ['comment-textless-close-button', 'comment-close-button'].forEach(
+      closeButtonTestId => {
+        it('closes popover when clicking close buttons', () => {
+          const wrapper = renderForm({ acceptComments: true });
+
+          togglePopover(wrapper);
+          assert.isTrue(commentPopoverExists(wrapper));
+
+          wrapper
+            .find(`button[data-testid="${closeButtonTestId}"]`)
+            .simulate('click');
+          assert.isFalse(commentPopoverExists(wrapper));
+        });
+      }
+    );
+
+    it('focuses comment textarea when popover is opened', async () => {
+      const wrapper = renderForm({ acceptComments: true });
+      togglePopover(wrapper);
+
+      assert.equal(
+        document.activeElement,
+        wrapper.find('textarea').getDOMNode()
+      );
+    });
+
+    it('submits grade using internal popover submit button', async () => {
+      const wrapper = renderForm({ acceptComments: true });
+
+      await waitForGradeFetch(wrapper);
+      togglePopover(wrapper);
+
+      wrapper
+        .find('button[data-testid="comment-submit-button"]')
+        .simulate('click');
+
+      assert.calledWith(fakeGradingService.submitGrade, {
+        student: fakeStudent,
+        grade: 1,
+        comment: 'Good job!',
```

Comment:
I think it would be good to have a test case which both inputs a comment and verifies that it was submitted as part of the same test, since that is the core workflow.

Reply:
Good idea.

I also wanted to test that no comment is sent as long as the feature is disabled, and forgot it. I will add both.